File: DataTransformAndAnalysis/python/AggregateByKey.py
from pyspark import SparkConf, SparkContext

conf = SparkConf().setAppName("Aggregate by key")
sc = SparkContext(conf=conf)

ordersRdd = sc.textFile("/user/cloudera/tableData/orders")
#Take order status as key and use anything for value(0 in this case) as we're using countByKey() action to calculate count
ordersMap = ordersRdd.map(lambda rec: (rec.split(",")[3],0))

#countByKey() example
#ordersMap.countByKey() #will return list
for i in ordersMap.countByKey().items():
	print i
	
#groupByKey() example
ordersMap = ordersRdd.map(lambda rec: (rec.split(",")[3],1))
ordersGroup = ordersMap.groupByKey()
ordersGroupCount = ordersGroup.map(lambda rec: (rec[0], sum(rec[1])))
for i in ordersGroupCount.collect():
	print i
	
#reduceByKey() example
#Input & Output type must be same. Combiner & Reducer logic is same
orderCount = ordersMap.reduceByKey(lambda x,y: x+y)
for i in orderCount.collect():
	print i

#combineByKey() example
#Input & Output type is same but combiner & reduce logic can differ
#TODO

	
#aggregateByKey() example
#Input & Output type can be different and combiner & reducer logic can also be different
#TODO

File: DataTransformAndAnalysis/scala/AggregateByKey.scala
/**
* sbt package
* spark-submit --class "AggregateByKey" --master local /home/cloudera/scala/target/scala-2.10/aggregatebykey-project_2.10-1.0.jar
**/
import org.apache.spark.SparkContext, org.apache.spark.SparkConf
import org.apache.spark.sql.hive.HiveContext

object AggregateByKey{
	def main(args: Array[String]){
	
		val conf=new SparkConf().setAppName("AggregateByKey")
		val sc = new SparkContext(conf)
		
		val orders = sc.textFile("/user/cloudera/tableData/orders")
		//countByKey()
		orders.map(rec => (rec.split(",")(3), rec)).countByKey().foreach(println)
		
		val ordersMap = orders.map(rec => (rec.split(",")(3), 1))
		//groupByKey()
		ordersMap.groupByKey().map(rec => (rec._1, rec._2.sum)).collect().foreach(println)
		
		//reduceByKey() - need to have same combine & reduce logic as well input/output datatype should be same
		ordersMap.reduceByKey((acc, value) => acc + value).collect().foreach(println)\
		
		//combineByKey() when tuple is 1 based. Can have different combine & reduce logic but input/output data type should be same
		val ordersMap = orders.map(rec => (rec.split(",")(3), 1))
		val ordersResult = ordersMap.combineByKey(value=>1, (acc : Int, value : Int)=> acc+value, (acc: Int, value: Int) => acc+value)
		ordersResult.collect().foreach(println)
		
		//combineByKey() when tuple is 0 based
		val ordersMap = orders.map(rec => (rec.split(",")(3), 0))
		val ordersResult = ordersMap.combineByKey(value=>1, (acc : Int, value : Int)=> acc+1, (acc: Int, value: Int) => acc+value)
		ordersResult.collect().foreach(println)
		
		//aggregateByKey() when tuple is 1 based - can have different combine & reduce logic. Can also have different input/output datatype
		val ordersMap = orders.map(rec => (rec.split(",")(3), 1))
		val ordersResult = ordersMap.aggregateByKey(0)((acc, value)=> acc+1, (acc, value) => acc+value)
		ordersResult.collect().foreach(println)
		
		//aggregateByKey() 
		val ordersMap = orders.map(rec => (rec.split(",")(3), rec))
		val ordersResult = ordersMap.aggregateByKey(0)((acc, value)=> acc+1, (acc, value) => acc+value)
		ordersResult.collect().foreach(println)
	}
}
